src/coreclr/hosts/inc/coreclrhost.h

```diff
@@ -47,6 +47,25 @@ CORECLR_HOSTING_API(coreclr_initialize,
             void** hostHandle,
             unsigned int* domainId);
 
+//
+// Type of the callback function called by the coreclr_get_error_info
+//
+typedef void (*error_info_callback)(const char* line, void* arg);
+
+//
+// Initialize the CoreCLR. Creates and starts CoreCLR host and creates an app domain
+//
+// Parameters:
+//  callBack                - callback that will be called for each line of the error info
+//  arg                     - argument to pass to the callback
+//
+// Returns:
+//  HRESULT indicating status of the operation. S_OK if the error info was successfully retrieved
+//
+CORECLR_HOSTING_API(coreclr_get_error_info,
+            error_info_callback callBack,
+            void* arg);
+
```

Comment:
Since this is to basically hook this up to the host tracing system, I would prefer it used the same mechanism we use inside host for this. Both `hostfxr` and `hostpolicy` expose "set error writer" entry points (`hostfxr_set_error_writer` and `corehost_set_error_writer`). It would feel "natural" to do the same for `coreclr`, so something like:
```C++
typedef void (*coreclr_error_writer_fn) (const pal::char_t *message);

CORECLR_HOSTING_API(coreclr_set_error_writer, coreclr_error_writer_fn error_writer);
```

The semantics is that the parent component (in this case `hostpolicy`) calls `set_error_writer` before doing anything else (right after LoadLibrary/dlopen) and the component (`coreclr`) in this case can call the call back function at any point in time it wants.

The components "propagate" these, so that if something called `set_error_writer` on `hostpolicy` it will propagate all errors into that writer, including those coming from child components (`coreclr).

This is super helpful (important) for native hosting scenarios, as sometimes the caller wants to hide/swallow errors, or redirect them. For example the comhost will buffer all errors and then propagate them using COM interfaces. Windows apps will buffer the errors and send them to EventLog (and also print out) and so on.

Reply:
The thing I don't like much about such approach is that:
* The output happens at "random" times (not synchronized with other possible output of the runtime)
* Not all E_FAIL necessarily have to lead to a runtime failure. We don't know that in advance, so we would log even the benign ones. I am not sure if that can happen now, but I can imagine such thing can be added anytime.
* Formatting the string needs to happen "inline". So basically when we detect E_FAIL (or any other HR that we decide to handle in the future), we would need to perform the (I assume) heavy weight lookup and string formatting even if the E_FAIL in the end turns out to be benign (handled gracefully). I guess it is not a big deal for the coreclr initialization, but it might be an issue if we decided to keep the writer on all the time instead of just during the coreclr_initialize call.

Reply:
Actually, I have realized that if we call it just at the end of coreclr_initialize, the concerns I have listed above are moot. And it would give us an extra ability to use it in future for extra error logging if needed. I'll convert to this model.

Reply:
@vitek-karas I have updated the change based on your suggested new API, can you please take a look if it matches your expectations?